decisions/adr-0007-vulcan-workflows.md

```diff
@@ -0,0 +1,60 @@
+
+We wish to unify the space of workflows in the data library, and integrate the normal work of data scientists in our group (and others) into the library. To do so, we will introduce a new Workflow data store in Vulcan. The data store will consist of two parts, a database model and a git tree. Workflow records will associate at the project level (i.e., each project will have a list of workflows). There will be no workflow author; the workflow belongs to the project.
+
+### Project workflows
+The Vulcan API will allow creating and listing project workflows. Each workflow can be also be browsed via the Vulcan UI. Workflows can still be run to produce Figures in Vulcan, which allow the workflow to be parameterized and run via the Vulcan UI. The Vulcan "steps" UI will be improved to monitor progress of a running workflow and show logging.
+
+The Vulcan API will also expose a git remote endpoint for each project workflow. When the workflow is created, it will be initialized to an empty tree or forked from an existing workflow. This is a normal git remote and can be pushed/pulled from as desired; the endpoint can possibly validate the push using an update hook.
+
+### Using Snakemake for workflows
+
+The format of a Vulcan workflow is a Snakemake workflow. This may include steps in python, R, etc. Workflows may define containers via a Snakefile and conda yaml. Vulcan will keep track of containers and re-use them across workflows.
+
+Inputs to the workflow can come from two places: 1) the data library, via https get, and 2) the cache/ folder associated with a Vulcan figure. Similarly, outputs can be: 1) writes to the data library (Magma or Metis) and 2) writes to the cache/ directory of a given folder
+
+Workflows will be executed by running Snakemake in the appropriate environment. Each job will contain a Snakefile, a config.yml, Vulcan inputs and an Etna token, and will generate output which will be copied into Vulcan's cache when the workflow completes or halts. Here we remain agnostic about how jobs will be dispatched.
+
+### UI Queries
+
+Currently the Vulcan orchestrator has special hooks for "ui query" steps. Since we cannot expect the Snakemake orchestator to understand this logic, we will instead separate these out into a new set of Vulcan endpoints, which allow a workflow to set a requirement for a UI query associated with a particular step in the workflow. To make use of this interface, a step in a Snakemake workflow which expects to output a UI query JSON (i.e., the JSON response from the user) will use a vulcan client interface to set a query requirement and exit, halting the workflow due to the unsatisfied output requirement. The Vulcan UI will then be able to determine an unresolved query and will show it to the user in the UI, making use of the existing query interfaces to resolve the query. This results in the response being written to the appropriate output location, and the Snakemake workflow resumes with the requirement satisfied.
```

Comment:
> a step in a Snakemake workflow which expects to output a UI query JSON (i.e., the JSON response from the user) will use a vulcan client interface to set a query requirement and exit.

Better if the run blockage can come from the other end -- reaching a step that requires **input** from UI query's output JSON.  There are cases currently where inputs for UI queries are prepped by an early step, yet the ui-query output is not needed until additional steps are completed.  Stopping the execution only at this stage might save a significant amount of users' time.

Perhaps getting to the point where a UI query could be presented could still trigger the UI to be presented to the user?  And execution would only be stopped if the user doesn't confirm UI selections before the workflow gets to a step that requires them?

Reply:
Currently there are empty "ui-query" steps which contain no code, but cause the UI to show the inputs to the step and massage through a query. I'm imagining that instead of being blank, this step would have some code calling a client to set the input requirement; the halting would thus occur in the same place in the workflow.